### PR TITLE
Further improve type annotations, reduce mypy whitelist

### PR DESCRIPTION
**Pull request by danieleades**

### CODE REVIEW

File: sphinx/writers/texinfo.py [hunk 1]
```diff
@@ -115,7 +115,7 @@ class TexinfoWriter(writers.Writer):
 
     settings_defaults: Dict = {}
 
-    output: str = None
+    output: Optional[str] = None  # type: ignore[assignment]
```
Comment (danieleades) [inline sphinx/writers/texinfo.py:118]: false positive. Probably an error in the associated docutils type stubs (i haven't checked).

File: sphinx/builders/latex/transforms.py [hunk 2]
```diff
@@ -94,13 +94,16 @@ def expand_show_urls(self) -> None:
     def get_docname_for_node(self, node: Node) -> str:
         while node:
             if isinstance(node, nodes.document):
-                return self.env.path2doc(node['source'])
+                doc = self.env.path2doc(node['source'])
+                if doc is None:
+                    raise ValueError("unreachable code")
+                return doc
```
Comment (danieleades) [inline sphinx/builders/latex/transforms.py]: @tk0miya can you please check you're happy with this approach here? is `path2doc` genuinely *never* `None` in this situation?

perhaps there's a better way to refactor this. this doesn't feel very 'clean' as is

**Comment by danieleades** [comment]
i'm not planning further changes in this PR @AA-Turner, @tk0miya

this will need careful review, as these are larger changes than previous 'typing' PRs (a good chunk of the codebase now has strict optional checking though!)

**Comment by AA-Turner** [comment]
Thanks @danieleades, as always. At some point please add yourself to AUTHORS, you've now made a series of fairly substantial typing-related improvements.

A

**Pull request merged**